include/ginkgo/core/solver/lower_trs.hpp

```diff
@@ -0,0 +1,296 @@
+    };                                                                         \
+    friend ::gko::solver::EnableDefaultLowerTrsFactory<                        \
+        _factory_name, _lower_trs, _parameters_name##_type, ValueType,         \
+        IndexType>;                                                            \
+                                                                               \
+private:                                                                       \
+    _parameters_name##_type _parameters_name##_;                               \
+                                                                               \
+public:                                                                        \
+    static_assert(true,                                                        \
+                  "This assert is used to counter the false positive extra "   \
+                  "semi-colon warnings")
+
+
+/**
+ * LowerTrs is the triangular solver which solves the system L x = b, when L is
+ * a lower triangular matrix. It works best when passing in a matrix in CSR
+ * format. If the matrix is not in CSR, then the generate step converts it into
+ * a CSR matrix. The generation fails if the matrix is not convertible to CSR.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indices
+ *
+ * @ingroup solvers
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class LowerTrs : public EnableLinOp<LowerTrs<ValueType, IndexType>>,
+                 public Preconditionable {
+    friend class EnableLinOp<LowerTrs>;
+    friend class EnablePolymorphicObject<LowerTrs, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+
+    /**
+     * Gets the system operator (CSR matrix) of the linear system.
+     *
+     * @return the system operator (CSR matrix)
+     */
+    std::shared_ptr<const matrix::Csr<ValueType, IndexType>> get_system_matrix()
+        const
+    {
+        return system_matrix_;
+    }
+
+    /**
+     * Gets the right hand side of the linear system.
+     *
+     * @return the right hand side
+     */
+    std::shared_ptr<const matrix::Dense<ValueType>> get_rhs() const
+    {
+        return b_;
+    }
+
+    /**
+     * Returns the preconditioner operator used by the solver.
+     *
+     * @return the preconditioner operator used by the solver
+     */
+    std::shared_ptr<const LinOp> get_preconditioner() const override
+    {
+        return preconditioner_;
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * Preconditioner factory.
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER(
+            preconditioner, nullptr);
+    };
+#define GKO_COMMA ,
+    GKO_ENABLE_LOWER_TRS_FACTORY(LowerTrs<ValueType GKO_COMMA IndexType>,
```

Comment:
Not sure why I did not think about that before, but you could simply do:
```suggestion
    GKO_ENABLE_LOWER_TRS_FACTORY(LowerTrs, parameters, Factory);
```
Without needing `GKO_COMMA` (since the class itself seems to know its own template parameter).